src/node_quic_session.cc

```diff
@@ -243,39 +247,58 @@ void QuicSession::AssociateCID(ngtcp2_cid* cid) {
   Socket()->AssociateCID(&id, &scid);
 }
 
+// Like the silent close, the immediate close must start with
+// the JavaScript side, first shutting down any existing
+// streams before entering the closing period. Unlike silent
+// close, however, all streams are closed using proper
+// STOP_SENDING and RESET_STREAM frames and a CONNECTION_CLOSE
+// frame is ultimately sent to the peer. This makes the
+// naming a bit of a misnomer in that the connection is
+// not immediately torn down, but is allowed to drain
+// properly per the QUIC spec description of "immediate close".
 void QuicSession::ImmediateClose() {
-  // Like the silent close, the immediate close must start with
-  // the JavaScript side, first shutting down any existing
-  // streams before entering the closing period. Unlike silent
-  // close, however, all streams are closed using proper
-  // STOP_SENDING and RESET_STREAM frames and a CONNECTION_CLOSE
-  // frame is ultimately sent to the peer. This makes the
-  // naming a bit of a misnomer in that the connection is
-  // not immediately torn down, but is allowed to drain
-  // properly per the QUIC spec description of "immediate close".
-  Debug(this, "Immediate close");
-  HandleScope scope(env()->isolate());
-  Local<Context> context = env()->context();
-  Context::Scope context_scope(context);
+  // Calling either ImmediateClose or SilentClose will cause
+  // the QUICSESSION_FLAG_CLOSING to be set. In either case,
+  // we should never re-enter ImmediateClose or SilentClose.
+  CHECK(!IsFlagSet(QUICSESSION_FLAG_CLOSING));
+  SetFlag(QUICSESSION_FLAG_CLOSING);
+
   QuicError last_error = GetLastError();
+  Debug(this, "Immediate close with code %" PRIu64 " (%s)",
+        last_error.code,
+        ErrorFamilyName(last_error.family));
+
+  HandleScope scope(env()->isolate());
+  Context::Scope context_scope(env()->context());
+
   Local<Value> argv[] = {
     Number::New(env()->isolate(), static_cast<double>(last_error.code)),
     Integer::New(env()->isolate(), last_error.family)
   };
+
+  // Grab a shared pointer to this to prevent the QuicSession
```

Comment:
Looks like this call is missed on line 994 https://github.com/nodejs/quic/blob/63b1b2f81e92b91ebdb777baad77a48fb715e640/src/node_quic_session.cc#L989-L995

Reply:
@trivikr ... please take a look again. The line numbers don't appear to be matching up so I'm not sure what you're referring to here.

Reply:
I searched for calls of `MakeCallback(` in `node_quic_session.cc` and noticed that there are 15 of them https://github.com/nodejs/quic/blob/master/src/node_quic_session.cc

And noticed that following line is added only on 14 of them in this PR
`std::shared_ptr<QuicSession> ptr(this->shared_from_this())`

Line number:
* nodejs:master is 994
* jasnell:iteration-13 is 1064
https://github.com/jasnell/quic/blob/3b8a9efce8640ce49365d74503528ed0d9b2c5fb/src/node_quic_session.cc#L1059-L1065

Reply:
Just did a quick scan through and verify that all of the MakeCallback calls in node_quic_session.cc include shared_ptr